components/feature/qr/src/main/java/mozilla/components/feature/qr/QrFeature.kt

```diff
@@ -0,0 +1,77 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.qr
+
+import android.Manifest.permission.CAMERA
+import android.content.Context
+import android.support.annotation.VisibleForTesting
+import android.support.v4.app.FragmentManager
+import mozilla.components.support.ktx.android.content.isPermissionGranted
+
+typealias OnScanResult = (result: String) -> Unit
+typealias OnNeedToRequestPermissions = (permissions: Array<String>) -> Unit
+
+/**
+ * Feature implementation that provides QR scanning functionality via the [QrFragment].
+ *
+ * @property applicationContext a reference to the application context.
+ * @property fragmentManager a reference to a [FragmentManager], used to start the [QrFragment].
+ * @property onScanResult a callback invoked with the result of the QR scan. The callback will
+ * always be invoked on the main thread.
+ * @property onNeedToRequestPermissions a callback invoked when permissions
+ * need to be requested before a QR scan can be performed. Once the request
+ * is completed, [onPermissionsResult] needs to be invoked.
+ */
+class QrFeature(
+    private val applicationContext: Context,
+    private val fragmentManager: FragmentManager,
+    private val onScanResult: OnScanResult = { },
+    private val onNeedToRequestPermissions: OnNeedToRequestPermissions = { }
+) {
+    private var containerViewId: Int = 0
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal val scanCompleteListener: QrFragment.OnScanCompleteListener = object : QrFragment.OnScanCompleteListener {
+        override fun onScanComplete(result: String) {
+            onScanResult(result)
+        }
+    }
+
+    /**
+     * Starts the QR scanner fragment and listens for scan results.
+     *
+     * @param containerViewId id of the container whose fragment(s) are to be
+     * replaced with the scanner fragment, defaults to [android.R.id.content].
+     * @return true if the scanner was started or false if permissions still
+     * need to be requested.
+     */
+    fun scan(containerViewId: Int = android.R.id.content): Boolean {
+        this.containerViewId = containerViewId
+
+        val fragment = QrFragment.newInstance(scanCompleteListener)
+
+        return if (applicationContext.isPermissionGranted(CAMERA)) {
+            fragmentManager.beginTransaction()
+                    .replace(containerViewId, fragment)
+                    .addToBackStack(null)
```

Comment:
The backstack is a monster that can cause a bunch of unexpected situations; especially when combined with other transactions that do not use the backstack. I wonder if we should use our `BackHandler` mechanism here and ask the app to forward those events to us?

Reply:
Switched to `BackHandler`!